### PR TITLE
feat(usb): implement USB printing via vendor_id and product_id

### PR DESCRIPTION
**Pull request by ipetrovbg**
# This PR

Adds support for USB printing via `vendor_id` and `product_id`

The implementation is heavily inspired by `escpos-rs` crate but tries to use a bit more idiomatic Rust.

*Please feel free to cancel the PR if this wasn't the intention of the `Usb` struct or for any other reason you think valid.*

#### How to use
```rust
    let product_id = 0xa700; // find the correct product_id for your printer
    let vendor_id = 0x0525; // find the correct vendor_id for your printer
    let usb = Usb::new(vendor_id, product_id)?;

    let mut printer = Printer::new(usb, None, None);

    printer
        .chain_feed(5)?
        .chain_font("C")?
        .chain_align("lt")?
        .chain_style("bu")?
        .chain_size(0, 0)?
        .chain_text("The quick brown fox jumps over the lazy dog")?
        .chain_barcode("12345678", "EAN8", "", "", 0, 0)?
        .chain_feed(5)?
        .chain_cut(false)?
        .flush()
```
76E987B1-1379-4F20-A16F-66A4055C9E08

### CODE REVIEW

File: src/device.rs [hunk 1]
```diff
@@ -74,3 +85,104 @@ impl<W: io::Write> io::Write for File<W> {
...
+    /// use escposify::device::Usb;
+    ///
+    /// let product_id = 0xa700;
+    /// let vendor_id = 0x0525;
+    /// let usb = Usb::new(vendor_id, product_id)?;
+    /// let mut printer = Printer::new(usb, None, None);
+    /// ```
+    pub fn new(vendor_id: u16, product_id: u16) -> io::Result<Usb> {
+        let context = Context::new().unwrap();
+        let devices = context.devices().unwrap();
```
Comment (TheWaWaR) [inline src/device.rs:103]: Please handle the errors here.

File: src/device.rs [hunk 2]
```diff
@@ -74,3 +85,104 @@ impl<W: io::Write> io::Write for File<W> {
...
+                                    _vendor_id: vendor_id,
+                                    _product_id: product_id,
+                                    connection: dvc,
+                                    endpoint,
+                                });
+                            }
+                            Err(e) => {
+                                return Err(io::Error::new(io::ErrorKind::Other, e));
+                            }
+                        };
```
Comment (TheWaWaR) [inline src/device.rs:163]:
```suggestion
                        return dvc
                          .claim_interface(0)
                          .map(|_| Usb {
                              _vendor_id: vendor_id,
                              _product_id: product_id,
                              connection: dvc,
                              endpoint,
                          }).map_err(|e| io::Error::new(io::ErrorKind::Other, e));
```

File: src/device.rs [hunk 3]
```diff
@@ -74,3 +85,104 @@ impl<W: io::Write> io::Write for File<W> {
...
+                    })?;
+
+                match device.open() {
+                    Ok(mut dvc) => {
+                        if let Ok(active) = dvc.kernel_driver_active(0) {
+                            if active {
+                                match dvc.detach_kernel_driver(0) {
+                                    Ok(_) => (),
+                                    Err(e) => {
+                                        println!("Error detaching kernel driver: {:?}", e);
```
Comment (TheWaWaR) [inline src/device.rs:139]: This line should be removed.

File: src/device.rs (hunk 1 above)
Comment (ipetrovbg) [reply src/device.rs:103]: 🤦

File: Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "escposify"
-version = "0.5.3"
+version = "0.6.3"
```
Comment (TheWaWaR) [inline Cargo.toml:3]:
```suggestion
version = "0.6.0"
```

File: src/device.rs [hunk 5]
```diff
@@ -74,3 +85,101 @@ impl<W: io::Write> io::Write for File<W> {
...
+                match device.open() {
+                    Ok(mut dvc) => {
+                        if let Ok(active) = dvc.kernel_driver_active(0) {
+                            if active {
+                                match dvc.detach_kernel_driver(0) {
+                                    Ok(_) => (),
+                                    Err(e) => {
+                                        return Err(io::Error::new(io::ErrorKind::Other, e));
+                                    }
+                                };
```
Comment (TheWaWaR) [inline src/device.rs:143]:
```rust
if let Err(e) = dvc.detach_kernel_driver(0) {
    return Err(io::Error::new(io::ErrorKind::Other, e));
}
```

**Review by TheWaWaR — APPROVED** [review]

**Comment by TheWaWaR** [comment]
Thanks

**Comment by ipetrovbg** [comment]
Would you be able to create a release or it will be created automatically?

**Comment by TheWaWaR** [comment]
Will create it manually.

**Comment by ipetrovbg** [comment]
Thanks!

**Comment by TheWaWaR** [comment]
Published

**Pull request merged**